File: core-js/src/solver/operations.ts
import { KinematicSystem } from '../system/KinematicSystem';
import { Vector3 } from '../math/Vector3';
import { Matrix4x4 } from '../math/Matrix4x4';

export interface Operation {
  type: 'align_node';
  effectorNode: string;
  targetNode: string;
  pivotNode: string;
  jointId: string;
  movingBodies: string[];
}

export function applyOperation(system: KinematicSystem, operation: Operation): void {
  if (operation.type === 'align_node') {
    const effector = system.nodes.get(operation.effectorNode);
    const target = system.nodes.get(operation.targetNode);
    const pivot = system.nodes.get(operation.pivotNode);
    const joint = system.joints.get(operation.jointId);

    if (!effector || !target || !pivot || !joint) {
      throw new Error(`Operation Error: Required elements not found.`);
    }

    const pPos = new Vector3(...pivot.absoluteTransform.getTranslation());
    const ePos = new Vector3(...effector.absoluteTransform.getTranslation());
    const tPos = new Vector3(...target.absoluteTransform.getTranslation());

    // Joint axis in world space
    const worldAxis = pivot.absoluteTransform.rotateVector(new Vector3(...joint.axis)).normalize();

    let step = 0;

    if (joint.type === 'revolute') {
      const pToE = ePos.sub(pPos);
      const pToT = tPos.sub(pPos);

      if (pToE.length() < 1e-6 || pToT.length() < 1e-6) return;

      const pToEn = pToE.normalize();
      const pToTn = pToT.normalize();

      const dot = Math.max(-1, Math.min(1, pToEn.dot(pToTn)));
      const angle = Math.acos(dot);

      if (angle < 1e-6) return;

      const cross = pToEn.cross(pToTn).normalize();
      const projection = cross.dot(worldAxis);
      
      // Step size (with some damping for stability)
      step = angle * projection * 0.5;

    } else if (joint.type === 'prismatic') {
      const eToT = tPos.sub(ePos);
      step = eToT.dot(worldAxis) * 0.5;
    }

    if (Math.abs(step) < 1e-8) return;

    // Apply limits
    const newValue = Math.max(joint.limits[0], Math.min(joint.limits[1], joint.value + step));
    const actualStep = newValue - joint.value;
    if (Math.abs(actualStep) < 1e-8) return;

    joint.value = newValue;

    // Calculate Delta Transform Matrix at Pivot
    // DeltaT = M_pivot * R_local_step * M_pivot_inv
    const localStepMat = new Matrix4x4();
    const [ax, ay, az] = joint.axis;
    if (joint.type === 'revolute') {
      if (ax === 1) localStepMat.rotateX(actualStep);
      else if (ay === 1) localStepMat.rotateY(actualStep);
      else localStepMat.rotateZ(actualStep);
    } else {
      localStepMat.translate(ax * actualStep, ay * actualStep, az * actualStep);
    }

    const pivotInv = pivot.absoluteTransform.clone().invert();
    const deltaT = pivot.absoluteTransform.clone().multiply(localStepMat).multiply(pivotInv);

    // Apply DeltaT to all moving bodies
    for (const bodyId of operation.movingBodies) {
      const body = system.bodies.get(bodyId);
      if (body) {
        body.transform = deltaT.clone().multiply(body.transform);
        body.updateNodes();
      }
    }

    return;
  }

  throw new Error(`Unsupported operation type: ${(operation as any).type}`);
}

File: core-py/src/wigglesolve/solver/operations.py
from __future__ import annotations
import math
from dataclasses import dataclass, field
from typing import List, Literal
from wigglesolve.math.vector3 import Vector3
from wigglesolve.math.matrix4x4 import Matrix4x4


@dataclass
class Operation:
    type: Literal["align_node"]
    effector_node: str
    target_node: str
    pivot_node: str
    joint_id: str
    moving_bodies: List[str]


def apply_operation(system, operation: Operation) -> None:
    if operation.type == "align_node":
        effector = system.nodes.get(operation.effector_node)
        target = system.nodes.get(operation.target_node)
        pivot = system.nodes.get(operation.pivot_node)
        joint = system.joints.get(operation.joint_id)

        if not effector or not target or not pivot or not joint:
            raise ValueError("Operation Error: Required elements not found.")

        p_pos = Vector3(*pivot.absolute_transform.get_translation())
        e_pos = Vector3(*effector.absolute_transform.get_translation())
        t_pos = Vector3(*target.absolute_transform.get_translation())

        world_axis = pivot.absolute_transform.rotate_vector(
            Vector3(*joint.axis)
        ).normalize()

        step = 0.0

        if joint.type == "revolute":
            p_to_e = e_pos.sub(p_pos)
            p_to_t = t_pos.sub(p_pos)

            if p_to_e.length() < 1e-6 or p_to_t.length() < 1e-6:
                return

            p_to_en = p_to_e.normalize()
            p_to_tn = p_to_t.normalize()

            dot = max(-1.0, min(1.0, p_to_en.dot(p_to_tn)))
            angle = math.acos(dot)

            if angle < 1e-6:
                return

            cross = p_to_en.cross(p_to_tn).normalize()
            projection = cross.dot(world_axis)
            step = angle * projection * 0.5

        elif joint.type == "prismatic":
            e_to_t = t_pos.sub(e_pos)
            step = e_to_t.dot(world_axis) * 0.5

        if abs(step) < 1e-8:
            return

        new_value = max(joint.limits[0], min(joint.limits[1], joint.value + step))
        actual_step = new_value - joint.value
        if abs(actual_step) < 1e-8:
            return

        joint.value = new_value

        local_step_mat = Matrix4x4()
        ax, ay, az = joint.axis
        if joint.type == "revolute":
            if ax == 1:
                local_step_mat = local_step_mat.rotate_x(actual_step)
            elif ay == 1:
                local_step_mat = local_step_mat.rotate_y(actual_step)
            else:
                local_step_mat = local_step_mat.rotate_z(actual_step)
        else:
            local_step_mat = local_step_mat.translate(ax * actual_step, ay * actual_step, az * actual_step)

        pivot_inv = pivot.absolute_transform.invert()
        delta_t = pivot.absolute_transform.multiply(local_step_mat).multiply(pivot_inv)

        for body_id in operation.moving_bodies:
            body = system.bodies.get(body_id)
            if body:
                body.transform = delta_t.multiply(body.transform)
                body.update_nodes()

        return

    raise ValueError(f"Unsupported operation type: {operation.type}")
